refactor: replaced all whitelist and blacklist occurences by allowlis…

byhlel opened the pull request:
…t and blocklist

We want to demonstrate our openness and inclusiveness in our code

Resolves #5275

<!-- Thanks for submitting a pull request for Terasology! :-)
Please fill in some details about the PR, below.
If the PR contains source code please make sure to run Checkstyle on it first.
If you add unit tests we'll love you forever! 

You might also want to read "How to Work on a PR Efficiently":
https://github.com/MovingBlocks/Terasology/wiki/How-to-Work-on-a-PR-Efficiently
-->

### Contains

Resolves #5275 

### How to test

Everything works as before

### Outstanding before merging

byhlel committed:
refactor: replaced all whitelist and blacklist occurences by allowlist and blocklist

We want to demonstrate our openness and inclusiveness in our code

Resolves #5275

Cervator commented:
This is eerily reminding me of #5276 - my apologies @byhlel for thinking you may _also_ be a bot or some sort of AI-assisted PR generator, but I'd like to ask for some sort of sign of humanity before considering this PR. 

Like just saying hello and what brought you here, and confirming that you've actually been able to run the game from source with the updates in place?

byhlel commented:
Hey,
I've just had some free time so I've been running around on good first issue.
I launched the game on a VM but im not on the best setup currently so I'd actually some review on this.
I'll try to rerun the game on another VM soon if no-one has time by then.

Also checking my profile would have reassured you on the latter.
Have a good one!

Cervator commented:
Oh phiew then - thanks for the quick reply, just a weird timing issue with the other PR and the barebones PR description. Maybe a tool that partially auto-created the PR out of a commit message or something?

I'd also be curious how you ran it on a VM, when trying in the past here that has gone ... poorly :-)

Any VM-induced performance issues aside so long as it builds and runs that's probably enough to call it validated 👍 

Although, considering that a rename of config files is included for a headless server run, that might be worth a quick check as well. And I wonder if we had any part of that automated for the test server setup / stored game server images we use 🤔 

This shouldn't be too bad to try out though, so maybe myself or another maintainer can check that soon. I'm on parental leave and have pretty minimal time, just spotted this by chance. I'll try to keep this in a tab near the top of my browser and see if I can get to it :-)

Much appreciated!

byhlel commented:
No I just like to keep my commits as concise as possible and respecting the repo's guideline.
So unfortunately (or fortunately) no tools neither to commit nor PR except from maybe lazyness to describe further the PR which was to me straightforward.

![image](https://github.com/user-attachments/assets/cc1a2547-a1c8-4642-a6ad-0fedba83e4fc)

Seems fine to me but further review to make sure nothing was accidently broken might be relevant here.

BenjaminAmos commented on .gitignore:
```diff
@@ -94,9 +94,9 @@ config/metrics/
-blacklist.json
+# Ignore Allowlist and Denylist files
+allowlist.json
+denylist.json
```
Can we please retain the old flle names in the gitignore so that they are not accidentally committed from older workspaces in future?

byhlel replied:
applied

BenjaminAmos commented on docs/Modding-API.md:
```diff
@@ -1,13 +1,13 @@
 =================
 
-Terasology's engine uses a whitelisting approach to expose an API for modules using two primary methods and a rarely needed third one:
+Terasology's engine uses allowlisting approach to expose an API for modules using two primary methods and a rarely needed third one:
```
```suggestion
Terasology's engine uses an allowlisting approach to expose an API for modules using two primary methods and a rarely needed third one:
```

BenjaminAmos replied:
Or alternatively:
```suggestion
Terasology's engine uses allowlisting to expose an API for modules using two primary methods and a rarely needed third one:
```

byhlel replied:
applied

BenjaminAmos commented on engine/src/main/java/org/terasology/engine/network/internal/ServerConnectListManager.java:
```diff
@@ -28,14 +28,14 @@ public class ServerConnectListManager {
-        blacklistPath = PathManager.getInstance().getHomePath().resolve("blacklist.json");
-        whitelistPath = PathManager.getInstance().getHomePath().resolve("whitelist.json");
+        denylistPath = PathManager.getInstance().getHomePath().resolve("denylist.json");
+        allowlistPath = PathManager.getInstance().getHomePath().resolve("allowlist.json");
```
Since this is likely to a breaking change otherwise, can we continue to check for the old file names if the newer ones do not exist?

byhlel replied:
applied

BenjaminAmos commented on engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/SelectGameScreen.java:
```diff
@@ -149,20 +149,20 @@ public void onOpened() {
-            Path blacklistPath = PathManager.getInstance().getHomePath().resolve("blacklist.json");
-            Path whitelistPath = PathManager.getInstance().getHomePath().resolve("whitelist.json");
-            if (!Files.exists(blacklistPath)) {
+            Path denylistPath = PathManager.getInstance().getHomePath().resolve("denylist.json");
+            Path allowlistPath = PathManager.getInstance().getHomePath().resolve("allowlist.json");
+            if (!Files.exists(denylistPath)) {
```
Could you please check for the presence of the old file names if the new ones do not exist?

byhlel replied:
applied

BenjaminAmos reviewed (changes requested):
Thanks for doing this. We have sadly had pull requests that were very suspect in the recent past but they were clearly so (like #5264) and nothing like this. We are quite happy to accept quality contributions from anyone.

Ordinarily though, renaming APIs would be a breaking change that would need a strong justification to warrant large-scale changes to our module ecosystem.

The changes here, however, appear minimal and only affect internal APIs, so I see no harm in accepting it, so long as backwards compatiblity is maintained. Currently with these changes `whitelist.json` and `blacklist.json` will no longer be read. If you could change this to accept both old and file names still then it should work well.

byhlel committed:
refactor: allow-deny lists refactor also allows for backwards compatibility

Code strucutre is different in SelectGameScreen and ServerConnectListManager
due to compiling error otherwise.
We check if path exist then we set the value if it does,
otherwise it applies "previous" behaviour (meaning in case if even blacklist or whitelist)
do not exist it sets the file to null.

So no different behaviour is to be expected.

byhlel commented:
Yes thank you for pointing this out, hadn't thought much of backwards compatibility!

jdrueckert committed:
chore: fix checkstyle RightCurlyCheck errors

BenjaminAmos commented on engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/SelectGameScreen.java:
```diff
@@ -149,20 +149,27 @@ public void onOpened() {
+            if (!Files.exists(denylistPath)) {
+                denylistPath = PathManager.getInstance().getHomePath().resolve("blacklist.json");
+            }
+            
+            if (!Files.exists(allowlistPath)) {
+                allowlistPath = PathManager.getInstance().getHomePath().resolve("whitelist.json");
+            }
+            if (!Files.exists(denylistPath)) {
                 try {
-                    Files.createFile(blacklistPath);
+                    Files.createFile(denylistPath);
                 } catch (IOException e) {
-                    logger.error("IO Exception on blacklist generation", e);
+                    logger.error("IO Exception on denylist generation", e);
                 }
             }
-            if (!Files.exists(whitelistPath)) {
+            if (!Files.exists(allowlistPath)) {
                 try {
-                    Files.createFile(whitelistPath);
+                    Files.createFile(allowlistPath);
                 } catch (IOException e) {
-                    logger.error("IO Exception on whitelist generation", e);
+                    logger.error("IO Exception on allowlist generation", e);
                 }
             }
         }
```
As I understand this logic, it is not quite correct. I was not paying full attention last time to this area. Now it creates `blacklist.json` and `whitelist.json` files if no files are present, I think. It should never create new `blacklist.json` and `whitelist.json` files, so if those are present then the check for the corresponding new names can just be skipped.

All that this code should be doing is creating empty files if they do not exist, rather than trying to load them in. As such, it should not be creating files with the old names and should skip creating files with the new names if those old-named files are already present.

BenjaminAmos commented on engine/src/main/java/org/terasology/engine/network/internal/ServerConnectListManager.java:
```diff
@@ -28,14 +28,24 @@ public class ServerConnectListManager {
     public ServerConnectListManager(Context context) {
-        blacklistPath = PathManager.getInstance().getHomePath().resolve("blacklist.json");
-        whitelistPath = PathManager.getInstance().getHomePath().resolve("whitelist.json");
+        // although this seems redundant, compiler wouldn't accept assigning then checking
+        if (Files.exists(PathManager.getInstance().getHomePath().resolve("denylist.json"))) {
+            denylistPath = PathManager.getInstance().getHomePath().resolve("denylist.json");
+        } else {
+            denylistPath = PathManager.getInstance().getHomePath().resolve("blacklist.json");
+        }
+        // although this seems redundant, compiler wouldn't accept assigning then checking
+        if (Files.exists(PathManager.getInstance().getHomePath().resolve("allowlist.json"))) {
+            allowlistPath = PathManager.getInstance().getHomePath().resolve("allowlist.json");
+        } else {
+            allowlistPath = PathManager.getInstance().getHomePath().resolve("whitelist.json");
+        }
```
```suggestion
        // Check for old naming first to retain compatiblity, otherwise use new naming by default.
        Path potentialDenylistPath = PathManager.getInstance().getHomePath().resolve("blacklist.json");
        Path potentialAllowlistPath = PathManager.getInstance().getHomePath().resolve("whitelist.json");
        if (!Files.exists(potentialDenylistPath)) {
            potentialDenylistPath = PathManager.getInstance().getHomePath().resolve("denylist.json");
        }
        if (!Files.exists(potentialAllowlistPath)) {
            potentialAllowlistPath = PathManager.getInstance().getHomePath().resolve("allowlist.json");
        }
        denylistPath = potentialDenylistPath;
        allowlistPath = potentialAllowlistPath;
```

`denylistPath` and `allowlistPath` are final variables, so they can only be assigned once in the constructor. Using a temporary variable here should work (but I wrote this quickly so please test that it compiles and works).

Cervator commented:
Closing this as superseded by #5324 which just merged
